### DISCUSSION TITLE
How to get synchronous behavior in custom cypress commands

### DISCUSSION THREAD

**Discussion by BetaGrumm** [cypress-io/cypress#15635] [Questions and Help]
I'm working on a test for SSO login functionality for our application.
I worked through just about everything required to do this properly, and I essentially have it, except that one of my custom cypress commands that I use for "test prep" to delete any test user that might be lying around, runs asyncronously such that my create user command fails as the prior delete command hasn't finished yet. 

I think I've boiled this down to the fact that AWS.CognitoIdentityServiceProvider.deleteUser() has a call back in it and that is what's tripping things up.
The problem is, I'm not sure how to unravel this with in the context of Cypress' slightly special asynchronous behavior. 

Any insight would be much appreciated. 

Test Code:
```
describe('Standard SSO authentication', () => {
  before(() => {
    cy.log('Pre cleaning test user...');
    cy.deleteCognitoUser();
    cy.log('Creating test user...');
    cy.createCognitoUser();
    cy.log('Authenticating test user...');
    cy.authenticateCognitoUser();
  });

 // This is commented out to test my pre cleaning functions
 // after(() => {
 //   cy.log('Post cleaning test user...');
 //   cy.deleteCognitoUser();
 // });

  it('Can sign-in to SSO enabled site', () => {
    cy.log('Loading site...');
    cy.viewport(1024, 768);
    cy.visit('/profile#access_token=foo1234');
    cy.contains('Profile').should('exist');
    cy.contains('About Me').should('exist');
  });
});
```

Custom Cypress Commands:
```
import Amplify from '@aws-amplify/core';
import Auth from '@aws-amplify/auth';
import AWS from 'aws-sdk';

Amplify.configure({
  Auth: {
    region: 'us-west-2',
    userPoolId: Cypress.env('QA_AWS_COGNITO_USER_POOL_ID'),
    userPoolWebClientId: Cypress.env('QA_AWS_COGNITO_USER_POOL_WEB_CLIENT_ID')
  }
});

Cypress.Commands.add('createCognitoUser', () =>
  Auth.signUp({
    username: Cypress.env('CY_COGNITO_TEST_USER'),
    password: Cypress.env('CY_COGNITO_TEST_PASSWORD'),
    attributes: {
      email: Cypress.env('CY_COGNITO_TEST_USER'),
      'custom:subdomain': 'ssoTestSubdomain'
    }
  })
);

Cypress.Commands.add('authenticateCognitoUser', () =>
  Auth.signIn(Cypress.env('CY_COGNITO_TEST_USER'), Cypress.env('CY_COGNITO_TEST_PASSWORD'))
);

Cypress.Commands.add('deleteCognitoUser', () => {
  const CognitoIdentityServiceProvider = new AWS.CognitoIdentityServiceProvider({
    region: 'us-west-2'
  });
  // See if the test user exists by attempting to sign in at that user
  Auth.signIn(Cypress.env('CY_COGNITO_TEST_USER'), Cypress.env('CY_COGNITO_TEST_PASSWORD'))
    .then(() => Auth.currentSession())
    .then(({ accessToken }) =>
      CognitoIdentityServiceProvider.deleteUser(
        {
          AccessToken: accessToken.jwtToken
        },
        (err, data) =>
          err ? console.log('deleteUser error: ', err) : console.log('Test user deleted.')
      )
    )
    .catch(e => {
      console.log('Caught deleteCognitoUser message: ', e);
    });
});
```

Again, all of this works, the issue is, cypress does not wait for deleteCognitoUser() to finish before it moves on which means, if there ever is a user existing at the beginning of a test, then the create user function will fail as the user already exists, and the test will fail even though the delete function will eventually finish and the user will have been deleted. 

Thanks again, for any clarity on how to get cypress to wait for what is admittedly a strange custom function.

**Comment by herecydev** [comment]
I would use `cy.task` (https://docs.cypress.io/api/commands/task) to move these sort of things into node and not be running in the browser. That will wait for your promise (make sure you return it or async/await); it also separates your test assertions and commands from those more cleanup related tasks.
